# Part 1. Equation of a Slime

How many late days are you using for this assignment? 1

In [20]:
# Imports section
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score, KFold
from sklearn.preprocessing import PolynomialFeatures

## 1. Loading the dataset

In [21]:
# Using pandas load the dataset
data = pd.read_csv("science_data_large.csv")

# Output the first 15 rows of the data
print('First 15 rows of the data:')
print(data.head(15))
print()

# Display a summary of the table information (number of datapoints, etc.)
print('Summary of the table information:')
print(data.info())

First 15 rows of the data:
    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04

Summary of the table information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols 

## 2. Splitting the dataset

In [22]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = data[['Temperature °C', 'Mols KCL']]
y = data['Size nm^3']

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

## 3. Perform a Linear Regression

In [23]:
# Use sklearn to train a model on the training set
model = LinearRegression()
model.fit(X_train, y_train)

# Create a sample datapoint and predict the output of that sample with the trained model
sample_datapoint = pd.DataFrame([[200, 350]], columns=['Temperature °C', 'Mols KCL'])
predicted_output = model.predict(sample_datapoint)
print(f'Predicted output for the sample datapoint: {predicted_output[0]:.5g}')
print()

# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means
score = model.score(X_test, y_test)
print(f'Model score: {score * 100:.5g}%')

# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
coefficients = model.coef_
intercept = model.intercept_
print('Model coefficients:', [f'{c:.5g}' for c in coefficients])
print(f'Model intercept: {intercept:.5g}')

Predicted output for the sample datapoint: 1.2528e+05

Model score: 85.525%
Model coefficients: ['866.15', '1032.7']
Model intercept: -4.0939e+05


Write the linear equation of a slime: (example equation: $E = mc^2$)

- $h(x) = 866.15x_1 + 1032.7x_2 - 409390$, where $x_1$ and $x_2$ represent the temperature (°C) and mols KCL, respectively.

Report on score and explain meaning:

- The model score, 85.525%, indicates that the model fits the data quite well. Hence, we can reliably use this linear regression model to predict future slime growth!

## 4. Use Cross Validation

In [24]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42
cv_strategy = KFold(n_splits=5, random_state=42, shuffle=True)
cv_scores = cross_val_score(model, X, y, cv=cv_strategy)

# Report on their finding and their significance
print('Cross validation scores:', [f'{score*100:.5g}%' for score in cv_scores])
print(f'Mean cross validation score: {cv_scores.mean()*100:.5g}%')

Cross validation scores: ['86.152%', '82.742%', '87.195%', '88.166%', '85.609%']
Mean cross validation score: 85.973%


Write findings here:

- The 5 cross validation scores across all splits indicate that the linear regression model is reliable for predications of unseen data. With not much perturbation between scores, we can confirm that the model is stable. The mean cross validation score, 85.973%, further validates that the model can be confidently used for predicting slime growth under varying conditions of temperature and KCl concentration!

## 5. Using Polynomial Regression

In [25]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly = PolynomialFeatures(degree=2, include_bias=False)
X_poly = poly.fit_transform(X)

# Perform k-fold cross validation (as above)
cv_strategy = KFold(n_splits=5, random_state=42, shuffle=True)
poly_model = LinearRegression()
cv_scores_poly = cross_val_score(poly_model, X_poly, y, cv=cv_strategy)

print(f'Mean Polynomial Regression cross-validation score: {cv_scores_poly.mean()*100:.5g}%')
print()

poly_model.fit(X_poly, y) # Fit the polynomial regression model

# Report on the metrics and output the resultant equation as you did in Part 3.
coefficients_poly = poly_model.coef_
intercept_poly = poly_model.intercept_
print('Polynomial model coefficients:', [f'{c:.5g}' for c in coefficients_poly])
print(f'Polynomial model intercept: {intercept_poly:.5g}')

Mean Polynomial Regression cross-validation score: 100%

Polynomial model coefficients: ['12', '-1.2311e-07', '-1.0593e-11', '2', '0.028571']
Polynomial model intercept: 1.6575e-05


Write the polynomial equation of a slime: (example equation: $E = mc^2$)

- $h(x) = (1.6575 \times 10^{-5}) + 12x_1 - (1.2311 \times 10^{-7})x_2 - (1.0593 \times 10^{-11})x_1^2 + 2x_1x_2 + 0.028571x_2^2$ 
    - where $x_1$ and $x_2$ represent the temperature (°C) and mols KCL, respectively.

Report on the score and interpret:

- The polynomial regression achieved a mean score of 100%, indicating a perfect fit. Hence, the polynomial regression model captures all variability in the slime growth data, outpreforming the linear model!

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [26]:
# Further imports
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier

In [27]:
# Load the dataset. Then train and evaluate the classification models.
ckd_data = pd.read_csv("ckd_feature_subset.csv")
x = ckd_data[['age', 'bp', 'wbcc', 'appet_poor', 'appet_good', 'rbcc']]
y = ckd_data['Target_ckd']

cv_strategy = KFold(n_splits=5, random_state=42, shuffle=True)
log_reg_model = LogisticRegression(random_state=42, max_iter=1000)
svc_model = SVC(random_state=42)
knn_model = KNeighborsClassifier()
nn_model = MLPClassifier(random_state=42, max_iter=1000)

models = {
    "Logistic Regression": log_reg_model,
    "Support Vector Machine": svc_model,
    "k-Nearest Neighbors": knn_model,
    "Neural Network": nn_model
}

results = {}

for name, model in models.items():
    scores = cross_val_score(model, x, y, cv=cv_strategy)
    results[name] = {"Mean Accuracy": f"{scores.mean()*100:.5g}", "Std Accuracy": f"{scores.std()*100:.5g}"}
    print(f"{name} - Accuracy scores: {[f'{s*100:.5g}' for s in scores]}")
    print(f"{name} - Mean Accuracy: {scores.mean()*100:.5g}, Std: {scores.std()*100:.5g}")
    print("------------------------------------------------")

Logistic Regression - Accuracy scores: ['80.645', '93.548', '77.419', '83.333', '93.333']
Logistic Regression - Mean Accuracy: 85.656, Std: 6.6269
------------------------------------------------
Support Vector Machine - Accuracy scores: ['90.323', '100', '87.097', '90', '96.667']
Support Vector Machine - Mean Accuracy: 92.817, Std: 4.7601
------------------------------------------------
k-Nearest Neighbors - Accuracy scores: ['93.548', '100', '87.097', '86.667', '96.667']
k-Nearest Neighbors - Mean Accuracy: 92.796, Std: 5.244
------------------------------------------------
Neural Network - Accuracy scores: ['90.323', '96.774', '87.097', '96.667', '96.667']
Neural Network - Mean Accuracy: 93.505, Std: 4.0466
------------------------------------------------


## Results and Conclusion for Classification Experiments

In [28]:
results_df = pd.DataFrame(results).T
results_df = results_df.sort_index()  
print()
print("Summary of Classification Model Performance:")
print(results_df)


Summary of Classification Model Performance:
                       Mean Accuracy Std Accuracy
Logistic Regression           85.656       6.6269
Neural Network                93.505       4.0466
Support Vector Machine        92.817       4.7601
k-Nearest Neighbors           92.796        5.244


From the table above, we observe that:
1. Neural Network is the highest accuracy model.
2. The SVM & KNN models had very similar performance.
3. Logistic Regression is behind all the other models, although it still is not bad with $85.656$% accuracy.

- Hence, we can conclude that the Neural Network model is the best performing model! Similar to part 1, the results indicate that a non-linear approach has higher accuracy and stability.

## Results and Conclusion for Neural Network Experiments

In [29]:
# Experiments with Neural Network.
nn_configurations = {
    "NN 1": MLPClassifier(
        hidden_layer_sizes=(10,),  # Single hidden layer with 10 neurons
        activation='relu',
        random_state=42,
        max_iter=10000
    ),
    "NN 2": MLPClassifier(
        hidden_layer_sizes=(20, 10),  
        activation='relu',
        random_state=42,
        max_iter=10000
    ),
    "NN 3": MLPClassifier(
        hidden_layer_sizes=(50, 50),  
        activation='relu',
        random_state=42,
        max_iter=10000
    ),

    "NN 4": MLPClassifier(
        hidden_layer_sizes=(500, 100, 50, 10),  
        activation='relu',
        random_state=42,
        max_iter=10000
    ),

    "NN 5": MLPClassifier(
        hidden_layer_sizes=(1000, 500),  
        activation='relu',
        random_state=42,
        max_iter=10000
    ),
}

nn_results = {}
for config_name, nn_model in nn_configurations.items():
    scores = cross_val_score(nn_model, x, y, cv=cv_strategy)
    nn_results[config_name] = {
        "Mean Accuracy (%)": f"{scores.mean()*100:.5g}",
        "Std Accuracy (%)": f"{scores.std()*100:.5g}"
    }

# Display the results
nn_results_df = pd.DataFrame(nn_results).T
print("Neural Network Experiment Results:")
print(nn_results_df)

Neural Network Experiment Results:
     Mean Accuracy (%) Std Accuracy (%)
NN 1            92.172           5.6307
NN 2            94.817           5.2335
NN 3            95.462           4.3696
NN 4            96.753            3.534
NN 5            96.753            3.534


From the table above, we observe that:
1. Neural Network configurations 4 and 5 achieved the highest accuracy.

- Hence, we can conclude that a Neural Network's accuracy increases as it gets more complex. This is evident because of the performance of a neural network as we increase the number and size of hidden layers. Moreover, notice that the accuracy seems to approach a threshhold, of $96.753$%, as the networks get more complex.

In [ ]:
# Test